### DISCUSSION TITLE
Other than validator, Is there a way to apply implicit to-`str` conversion?

### DISCUSSION THREAD

**Discussion by duongdominhchau** [pydantic/pydantic#7202] [Question]
> Strings are accepted as-is, `bytes` and `bytearray` are converted using `v.decode()`, enums inheriting from `str` are converted using `v.value`, and **all other types cause an error**

(Emphasis mine)

My problem is: my database uses both `int` and `UUID` as ID type, but I don't want to expose the raw type so I declared the `id` as a `str` in my Pydantic model. This caused the error

> Input should be a valid string [type=string_type, input_value=1, input_type=int]

Example code:

```py
from dataclasses import dataclass

from pydantic import BaseModel, ConfigDict, field_validator


class MyModel(BaseModel):
    id: str
    model_config = ConfigDict(from_attributes=True)


@dataclass
class SomethingElse:
    id: int


o = SomethingElse(id=1)
print(MyModel.model_validate(o).model_dump_json())
```

I thought of using a `before` validator to do this, but it feels weird having to define a method for each attribute like this. Is there another way to relax the constraint so `str()` is called for whatever passed into this model constructor?

**Comment by harunyasar** [comment]
```python
import uuid
from typing import Generic, TypeVar

from pydantic import BaseModel, ConfigDict, field_validator

T = TypeVar("T", int, uuid.UUID)


class MyModel(Generic[T], BaseModel):
    id: T
    model_config = ConfigDict(from_attributes=True)


o1 = MyModel(id=1)
o2 = MyModel(id=uuid.uuid4())


print(MyModel.model_validate(o1).model_dump_json())
print(MyModel.model_validate(o2).model_dump_json())

# {"id":1}
# {"id":"f7513692-9ff7-4b46-9745-1820008e834c"}
```

It seems to me that your problem can be solved using Generic types. If you are expecting an integer input to be serialized in string type as well, you could create a base model with a `@field_serializer` by inherting from it.

**Reply by duongdominhchau** [reply]
What I have in my mind is a bit different though, I want to discard the knowledge about specific type of `id` so everyone using this model must treat it as `str` and not `int | UUID`, so the conversion needs to happen on the way in, not just out. However, maybe my original line of reasoning is wrong already, let me think more about this for a while. Thank you for this alternative solution, I didn't have even a vague thought of that when I wrote this question.

**Reply by harunyasar** [reply]
Now I understand what you mean. You would like to be able to pass `UUID` object or an `int` to a string field defined in your Pydantic model. Am I right?

```python
m = MyModel(id=1081)  # MyModel(id='1081')
m = MyModel(id=uuid.uuid4())  # MyModel(id='59810243-6cdf-4936-931e-d96e2e135290')
```

I really don't know how the default type coercion logic during instantiation could be overridden.

**Reply by duongdominhchau** [reply]
Yes, basically what I want is something like this:

```py
from dataclasses import dataclass
from typing import Any

from pydantic import BaseModel, ConfigDict, field_validator


class MyModel(BaseModel):
    id: str
    model_config = ConfigDict(from_attributes=True)

    @field_validator("id", mode="before")
    @classmethod
    def _id_to_str(cls, value: Any) -> str:
        return str(value)


@dataclass
class SomethingElse:
    id: int


o = SomethingElse(id=1)
print(MyModel.model_validate(o).model_dump_json()) # => {"id":"1"}
```

but without having to define custom validator for each field needing that behavior.

**Reply by harunyasar** [reply]
I don't know if you have any other way but to do this with `@field_validaor`. If you ask me, the best way would be to put it in a main model and use it within the sub models.

**Reply by duongdominhchau** [reply]
I see, thank you for the ideas. Closing this then.
